File: client/js/src/video/consumer.ts
/**
 * Consumer client for receiving video streams in LeRobot Arena
 */

import { VideoClientCore } from './core.js';
import type {
  WebSocketMessage,
  FrameUpdateMessage,
  VideoConfigUpdateMessage,
  StreamStartedMessage,
  StreamStoppedMessage,
  RecoveryTriggeredMessage,
  StatusUpdateMessage,
  StreamStatsMessage,
  ClientOptions,
  WebRTCStats,
  FrameUpdateCallback,
  VideoConfigUpdateCallback,
  StreamStartedCallback,
  StreamStoppedCallback,
  RecoveryTriggeredCallback,
  StatusUpdateCallback,
  StreamStatsCallback,
  WebRTCOfferMessage,
  WebRTCIceMessage,
} from './types.js';

export class VideoConsumer extends VideoClientCore {
  // Event callbacks
  private onFrameUpdateCallback: FrameUpdateCallback | null = null;
  private onVideoConfigUpdateCallback: VideoConfigUpdateCallback | null = null;
  private onStreamStartedCallback: StreamStartedCallback | null = null;
  private onStreamStoppedCallback: StreamStoppedCallback | null = null;
  private onRecoveryTriggeredCallback: RecoveryTriggeredCallback | null = null;
  private onStatusUpdateCallback: StatusUpdateCallback | null = null;
  private onStreamStatsCallback: StreamStatsCallback | null = null;

  // ICE candidate queuing for proper timing
  private iceCandidateQueue: { candidate: RTCIceCandidate; fromProducer: string }[] = [];
  private hasRemoteDescription = false;

  constructor(baseUrl = 'http://localhost:8000', options: ClientOptions = {}) {
    super(baseUrl, options);
  }

  // ============= CONSUMER CONNECTION =============

  async connect(workspaceId: string, roomId: string, participantId?: string): Promise<boolean> {
    const connected = await this.connectToRoom(workspaceId, roomId, 'consumer', participantId);
    
    if (connected) {
      // Create peer connection immediately so we're ready for WebRTC offers
      console.info('🔧 Creating peer connection for consumer...');
      await this.startReceiving();
    }
    
    return connected;
  }

  // ============= CONSUMER METHODS =============

  async startReceiving(): Promise<void> {
    if (!this.connected) {
      throw new Error('Must be connected to start receiving');
    }

    // Reset WebRTC state
    this.hasRemoteDescription = false;
    this.iceCandidateQueue = [];

    // Create peer connection for receiving
    this.createPeerConnection();
    
    // Set up to receive remote stream
    if (this.peerConnection) {
      this.peerConnection.ontrack = (event: RTCTrackEvent) => {
        console.info('📺 Received remote track:', event.track.kind);
        this.remoteStream = event.streams[0] || null;
        this.emit('remoteStream', this.remoteStream);
        this.emit('streamReceived', this.remoteStream);
      };
    }
  }

  async stopReceiving(): Promise<void> {
    if (this.peerConnection) {
      this.peerConnection.close();
      this.peerConnection = null;
    }
    this.remoteStream = null;
    this.emit('streamStopped');
  }

  // ============= WEBRTC NEGOTIATION =============

  async handleWebRTCOffer(message: WebRTCOfferMessage): Promise<void> {
    try {
      console.info(`📥 Received WebRTC offer from producer ${message.from_producer}`);
      
      if (!this.peerConnection) {
        console.warn('No peer connection available to handle offer');
        return;
      }

      // Reset state for new offer
      this.hasRemoteDescription = false;
      this.iceCandidateQueue = [];

      // Set remote description (the offer)
      await this.setRemoteDescription(message.offer);
      this.hasRemoteDescription = true;
      
      // Process any queued ICE candidates now that we have remote description
      await this.processQueuedIceCandidates();
      
      // Create answer
      const answer = await this.createAnswer(message.offer);
      
      console.info(`📤 Sending WebRTC answer to producer ${message.from_producer}`);
      
      // Send answer back through server to producer
      if (this.workspaceId && this.roomId && this.participantId) {
        await this.sendWebRTCSignal(this.workspaceId, this.roomId, this.participantId, {
          type: 'answer',
          sdp: answer.sdp,
          target_producer: message.from_producer,
        } as Record<string, unknown>);
      }
      
      console.info('✅ WebRTC negotiation completed from consumer side');
    } catch (error) {
      console.error('Failed to handle WebRTC offer:', error);
      this.handleError(`Failed to handle WebRTC offer: ${error}`);
    }
  }

  private async handleWebRTCIce(message: WebRTCIceMessage): Promise<void> {
    if (!this.peerConnection) {
      console.warn('No peer connection available to handle ICE');
      return;
    }

    try {
      console.info(`📥 Received WebRTC ICE from producer ${message.from_producer}`);
      
      const candidate = new RTCIceCandidate(message.candidate);
      
      if (!this.hasRemoteDescription) {
        // Queue ICE candidate until we have remote description
        console.info(`🔄 Queuing ICE candidate from ${message.from_producer} (no remote description yet)`);
        this.iceCandidateQueue.push({
          candidate,
          fromProducer: message.from_producer || 'unknown'
        });
        return;
      }
      
      // Add ICE candidate to peer connection
      await this.peerConnection.addIceCandidate(candidate);
      
      console.info(`✅ WebRTC ICE handled from producer ${message.from_producer}`);
    } catch (error) {
      console.error(`Failed to handle WebRTC ICE from ${message.from_producer}:`, error);
      this.handleError(`Failed to handle WebRTC ICE: ${error}`);
    }
  }

  private async processQueuedIceCandidates(): Promise<void> {
    if (this.iceCandidateQueue.length === 0) {
      return;
    }

    console.info(`🔄 Processing ${this.iceCandidateQueue.length} queued ICE candidates`);
    
    for (const { candidate, fromProducer } of this.iceCandidateQueue) {
      try {
        if (this.peerConnection) {
          await this.peerConnection.addIceCandidate(candidate);
          console.info(`✅ Processed queued ICE candidate from ${fromProducer}`);
        }
      } catch (error) {
        console.error(`Failed to process queued ICE candidate from ${fromProducer}:`, error);
      }
    }
    
    // Clear the queue
    this.iceCandidateQueue = [];
  }

  // Override to add producer targeting for ICE candidates
  override createPeerConnection(): RTCPeerConnection {
    const config: RTCConfiguration = {
      iceServers: this.webrtcConfig.iceServers || [
        { urls: 'stun:stun.l.google.com:19302' }
      ]
    };

    this.peerConnection = new RTCPeerConnection(config);

    // Connection state changes
    this.peerConnection.onconnectionstatechange = () => {
      const state = this.peerConnection?.connectionState;
      console.info(`🔌 WebRTC connection state: ${state}`);
    };

    // ICE connection state
    this.peerConnection.oniceconnectionstatechange = () => {
      const state = this.peerConnection?.iceConnectionState;
      console.info(`🧊 ICE connection state: ${state}`);
    };

    // ICE candidate handling - send to producer
    this.peerConnection.onicecandidate = (event: RTCPeerConnectionIceEvent) => {
      if (event.candidate && this.workspaceId && this.roomId && this.participantId) {
        // Send ICE candidate to producer
        this.sendIceCandidateToProducer(event.candidate);
      }
    };

    // Handle remote stream
    this.peerConnection.ontrack = (event: RTCTrackEvent) => {
      console.info('📺 Received remote track:', event.track.kind);
      this.remoteStream = event.streams[0] || null;
      this.emit('remoteStream', this.remoteStream);
      this.emit('streamReceived', this.remoteStream);
    };

    return this.peerConnection;
  }

  private async sendIceCandidateToProducer(candidate: RTCIceCandidate): Promise<void> {
    if (!this.workspaceId || !this.roomId || !this.participantId) return;

    try {
      // Get room info to find the producer
      const roomInfo = await this.getRoomInfo(this.workspaceId, this.roomId);
      
      if (roomInfo.participants.producer) {
        await this.sendWebRTCSignal(this.workspaceId, this.roomId, this.participantId, {
          type: 'ice',
          candidate: candidate.toJSON(),
          target_producer: roomInfo.participants.producer,
        } as Record<string, unknown>);
      }
    } catch (error) {
      console.error('Failed to send ICE candidate to producer:', error);
    }
  }

  private async handleStreamStarted(message: StreamStartedMessage): Promise<void> {
    if (this.onStreamStartedCallback) {
      this.onStreamStartedCallback(message.config, message.participant_id);
    }
    this.emit('streamStarted', message.config, message.participant_id);

    console.info(`🚀 Stream started by producer ${message.participant_id}, ready to receive video`);
  }

  // ============= EVENT CALLBACKS =============

  onFrameUpdate(callback: FrameUpdateCallback): void {
    this.onFrameUpdateCallback = callback;
  }

  onVideoConfigUpdate(callback: VideoConfigUpdateCallback): void {
    this.onVideoConfigUpdateCallback = callback;
  }

  onStreamStarted(callback: StreamStartedCallback): void {
    this.onStreamStartedCallback = callback;
  }

  onStreamStopped(callback: StreamStoppedCallback): void {
    this.onStreamStoppedCallback = callback;
  }

  onRecoveryTriggered(callback: RecoveryTriggeredCallback): void {
    this.onRecoveryTriggeredCallback = callback;
  }

  onStatusUpdate(callback: StatusUpdateCallback): void {
    this.onStatusUpdateCallback = callback;
  }

  onStreamStats(callback: StreamStatsCallback): void {
    this.onStreamStatsCallback = callback;
  }

  // ============= MESSAGE HANDLING =============

  protected override handleRoleSpecificMessage(message: WebSocketMessage): void {
    switch (message.type) {
      case 'frame_update':
        this.handleFrameUpdate(message as FrameUpdateMessage);
        break;
      case 'video_config_update':
        this.handleVideoConfigUpdate(message as VideoConfigUpdateMessage);
        break;
      case 'stream_started':
        this.handleStreamStarted(message as StreamStartedMessage);
        break;
      case 'stream_stopped':
        this.handleStreamStopped(message as StreamStoppedMessage);
        break;
      case 'recovery_triggered':
        this.handleRecoveryTriggered(message as RecoveryTriggeredMessage);
        break;
      case 'status_update':
        this.handleStatusUpdate(message as StatusUpdateMessage);
        break;
      case 'stream_stats':
        this.handleStreamStats(message as StreamStatsMessage);
        break;
      case 'participant_joined':
        console.info(`📥 Participant joined: ${message.participant_id} as ${message.role}`);
        break;
      case 'participant_left':
        console.info(`📤 Participant left: ${message.participant_id} (${message.role})`);
        break;
      case 'webrtc_offer':
        this.handleWebRTCOffer(message as WebRTCOfferMessage);
        break;
      case 'webrtc_answer':
        console.info('📨 Received WebRTC answer (consumer should not receive this)');
        break;
      case 'webrtc_ice':
        this.handleWebRTCIce(message as WebRTCIceMessage);
        break;
      case 'emergency_stop':
        console.warn(`🚨 Emergency stop: ${message.reason || 'Unknown reason'}`);
        this.handleError(`Emergency stop: ${message.reason || 'Unknown reason'}`);
        break;
      case 'error':
        console.error(`Server error: ${message.message}`);
        this.handleError(message.message);
        break;
      default:
        console.warn(`Unknown message type for consumer: ${message.type}`);
    }
  }

  private handleFrameUpdate(message: FrameUpdateMessage): void {
    if (this.onFrameUpdateCallback) {
      const frameData = {
        data: message.data,
        metadata: message.metadata
      };
      this.onFrameUpdateCallback(frameData);
    }
    this.emit('frameUpdate', message.data);
  }

  private handleVideoConfigUpdate(message: VideoConfigUpdateMessage): void {
    if (this.onVideoConfigUpdateCallback) {
      this.onVideoConfigUpdateCallback(message.config);
    }
    this.emit('videoConfigUpdate', message.config);
  }

  private handleStreamStopped(message: StreamStoppedMessage): void {
    if (this.onStreamStoppedCallback) {
      this.onStreamStoppedCallback(message.participant_id, message.reason);
    }
    this.emit('streamStopped', message.participant_id, message.reason);
  }

  private handleRecoveryTriggered(message: RecoveryTriggeredMessage): void {
    if (this.onRecoveryTriggeredCallback) {
      this.onRecoveryTriggeredCallback(message.policy, message.reason);
    }
    this.emit('recoveryTriggered', message.policy, message.reason);
  }

  private handleStatusUpdate(message: StatusUpdateMessage): void {
    if (this.onStatusUpdateCallback) {
      this.onStatusUpdateCallback(message.status, message.data);
    }
    this.emit('statusUpdate', message.status, message.data);
  }

  private handleStreamStats(message: StreamStatsMessage): void {
    if (this.onStreamStatsCallback) {
      this.onStreamStatsCallback(message.stats);
    }
    this.emit('streamStats', message.stats);
  }

  // ============= UTILITY METHODS =============

  /**
   * Create a consumer and automatically connect to a room
   */
  static async createAndConnect(
    workspaceId: string,
    roomId: string,
    baseUrl = 'http://localhost:8000',
    participantId?: string
  ): Promise<VideoConsumer> {
    const consumer = new VideoConsumer(baseUrl);
    const connected = await consumer.connect(workspaceId, roomId, participantId);
    
    if (!connected) {
      throw new Error('Failed to connect as video consumer');
    }
    
    return consumer;
  }

  /**
   * Get the video element for displaying the remote stream
   */
  attachToVideoElement(videoElement: HTMLVideoElement): void {
    if (this.remoteStream) {
      videoElement.srcObject = this.remoteStream;
    }
    
    // Listen for future stream updates
    this.on('remoteStream', (stream: MediaStream) => {
      videoElement.srcObject = stream;
    });
  }

  /**
   * Get current video statistics
   */
  async getVideoStats(): Promise<WebRTCStats | null> {
    const stats = await this.getStats();
    return stats;
  }
} 

File: client/python/src/transport_server_client/video/consumer.py
"""
Consumer client for receiving video streams in LeRobot Arena
"""

import asyncio
import logging
from typing import Any

from aiortc import RTCIceCandidate, RTCSessionDescription

from .core import VideoClientCore
from .types import (
    ClientOptions,
    FrameData,
    FrameUpdateCallback,
    ParticipantRole,
    RecoveryTriggeredCallback,
    StatusUpdateCallback,
    StreamStartedCallback,
    StreamStatsCallback,
    StreamStoppedCallback,
    VideoConfig,
    VideoConfigUpdateCallback,
    WebRTCStats,
)

logger = logging.getLogger(__name__)


class VideoConsumer(VideoClientCore):
    """Consumer client for receiving video streams in LeRobot Arena"""

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        options: ClientOptions | None = None,
    ):
        super().__init__(base_url, options)

        # Event callbacks
        self.on_frame_update_callback: FrameUpdateCallback | None = None
        self.on_video_config_update_callback: VideoConfigUpdateCallback | None = None
        self.on_stream_started_callback: StreamStartedCallback | None = None
        self.on_stream_stopped_callback: StreamStoppedCallback | None = None
        self.on_recovery_triggered_callback: RecoveryTriggeredCallback | None = None
        self.on_status_update_callback: StatusUpdateCallback | None = None
        self.on_stream_stats_callback: StreamStatsCallback | None = None

        # ICE candidate queuing for proper timing
        self.ice_candidate_queue: list[dict[str, Any]] = []
        self.has_remote_description = False

        # Frame monitoring for stream health
        self._last_frame_time: float | None = None
        self._frame_timeout_task: asyncio.Task | None = None
        self._monitoring_frames = False

    # ============= CONSUMER CONNECTION =============

    async def connect(
        self, workspace_id: str, room_id: str, participant_id: str | None = None
    ) -> bool:
        """Connect to a room as consumer"""
        # Create peer connection BEFORE connecting to avoid race condition
        logger.info("Creating peer connection for consumer...")
        self.create_peer_connection()

        # Add transceiver to receive video
        if self.peer_connection:
            self.peer_connection.addTransceiver("video", direction="recvonly")
            logger.info("Added video transceiver for consumer")

        # Now connect to room - we're ready for WebRTC offers
        connected = await self.connect_to_room(
            workspace_id, room_id, ParticipantRole.CONSUMER, participant_id
        )

        if connected:
            # Create peer connection immediately so we're ready for WebRTC offers
            logger.info("🔧 Consumer connected and ready for WebRTC offers")
            await self.start_receiving()

        return connected

    # ============= CONSUMER METHODS =============

    async def start_receiving(self) -> None:
        """Start receiving video stream"""
        if not self.connected:
            raise ValueError("Must be connected to start receiving")

        # Reset WebRTC state
        self.has_remote_description = False
        self.ice_candidate_queue = []

        # Create peer connection for receiving (if not already created)
        if not self.peer_connection:
            self.create_peer_connection()

            # Set up to receive remote stream
            if self.peer_connection:
                # Add transceiver to receive video
                self.peer_connection.addTransceiver("video", direction="recvonly")
                logger.info("Added video transceiver for consumer")
        else:
            logger.info("Peer connection already exists for consumer")

    async def stop_receiving(self) -> None:
        """Stop receiving video stream"""
        # Stop frame monitoring
        self._monitoring_frames = False
        if self._frame_timeout_task and not self._frame_timeout_task.done():
            self._frame_timeout_task.cancel()

        if self.peer_connection:
            await self.peer_connection.close()
            self.peer_connection = None
        self.remote_stream = None

    # ============= WEBRTC NEGOTIATION =============

    async def handle_webrtc_offer(
        self, offer_data: dict[str, Any], from_producer: str
    ) -> None:
        """Handle WebRTC offer from producer"""
        try:
            logger.info(f"📥 Received WebRTC offer from producer {from_producer}")

            # Check if we need to restart the connection (new offer from same producer)
            if self.peer_connection and self.has_remote_description:
                logger.info("🔄 Restarting connection for new stream...")
                await self._restart_connection_for_new_stream()

            if not self.peer_connection:
                logger.info("🔧 Creating new peer connection for offer...")
                self.create_peer_connection()

                # Add transceiver to receive video
                if self.peer_connection:
                    self.peer_connection.addTransceiver("video", direction="recvonly")
                    logger.info("Added video transceiver for new connection")

            # Reset state for new offer
            self.has_remote_description = False
            self.ice_candidate_queue = []

            # Set remote description (the offer)
            offer = RTCSessionDescription(
                sdp=offer_data["sdp"], type=offer_data["type"]
            )
            await self.set_remote_description(offer)
            self.has_remote_description = True

            # Process any queued ICE candidates now that we have remote description
            await self._process_queued_ice_candidates()

            # Create answer
            answer = await self.create_answer(offer)

            logger.info(f"📤 Sending WebRTC answer to producer {from_producer}")

            # Send answer back through server to producer
            if self.workspace_id and self.room_id and self.participant_id:
                await self.send_webrtc_signal(
                    self.workspace_id,
                    self.room_id,
                    self.participant_id,
                    {
                        "type": "answer",
                        "sdp": answer.sdp,
                        "target_producer": from_producer,
                    },
                )

            logger.info("✅ WebRTC negotiation completed from consumer side")
        except Exception as e:
            logger.error(f"Failed to handle WebRTC offer: {e}")
            if self.on_error_callback:
                self.on_error_callback(f"Failed to handle WebRTC offer: {e}")

    async def _restart_connection_for_new_stream(self) -> None:
        """Restart connection for new stream (called when getting new offer)"""
        try:
            logger.info("🔄 Restarting peer connection for new stream...")

            # Stop frame monitoring
            self._monitoring_frames = False
            if self._frame_timeout_task and not self._frame_timeout_task.done():
                self._frame_timeout_task.cancel()

            # Close existing peer connection
            if self.peer_connection:
                await self.peer_connection.close()
                self.peer_connection = None

            # Reset all WebRTC state
            self.remote_stream = None
            self.has_remote_description = False
            self.ice_candidate_queue = []
            self._last_frame_time = None

            logger.info("✅ Connection restart completed")

        except Exception as e:
            logger.error(f"❌ Error restarting connection: {e}")
            # Continue anyway - we'll try to create a new connection

    async def handle_webrtc_ice(
        self, ice_data: dict[str, Any], from_producer: str
    ) -> None:
        """Handle WebRTC ICE candidate from producer"""
        if not self.peer_connection:
            logger.warning("No peer connection available to handle ICE")
            return

        try:
            logger.info(f"📥 Received WebRTC ICE from producer {from_producer}")

            # Parse ICE candidate string and create RTCIceCandidate
            candidate_str = ice_data["candidate"]
            parts = candidate_str.split()

            if len(parts) >= 8:
                candidate = RTCIceCandidate(
                    component=int(parts[1]),
                    foundation=parts[0].split(":")[1],  # Remove "candidate:" prefix
                    ip=parts[4],
                    port=int(parts[5]),
                    priority=int(parts[3]),
                    protocol=parts[2],
                    type=parts[7],  # typ value
                    sdpMid=ice_data.get("sdpMid"),
                    sdpMLineIndex=ice_data.get("sdpMLineIndex"),
                )
            else:
                logger.warning(f"Invalid ICE candidate format: {candidate_str}")
                return

            if not self.has_remote_description:
                # Queue ICE candidate until we have remote description
                logger.info(
                    f"🔄 Queuing ICE candidate from {from_producer} (no remote description yet)"
                )
                self.ice_candidate_queue.append({
                    "candidate": candidate,
                    "from_producer": from_producer,
                })
                return

            # Add ICE candidate to peer connection
            await self.add_ice_candidate(candidate)

            logger.info(f"✅ WebRTC ICE handled from producer {from_producer}")
        except Exception as e:
            logger.error(f"Failed to handle WebRTC ICE from {from_producer}: {e}")
            if self.on_error_callback:
                self.on_error_callback(f"Failed to handle WebRTC ICE: {e}")

    async def _process_queued_ice_candidates(self) -> None:
        """Process all queued ICE candidates after remote description is set"""
        if not self.ice_candidate_queue:
            return

        logger.info(
            f"🔄 Processing {len(self.ice_candidate_queue)} queued ICE candidates"
        )

        for item in self.ice_candidate_queue:
            try:
                candidate = item["candidate"]
                from_producer = item["from_producer"]

                if self.peer_connection:
                    await self.peer_connection.addIceCandidate(candidate)
                    logger.info(
                        f"✅ Processed queued ICE candidate from {from_producer}"
                    )
            except Exception as e:
                logger.error(
                    f"Failed to process queued ICE candidate from {item.get('from_producer', 'unknown')}: {e}"
                )

        # Clear the queue
        self.ice_candidate_queue = []

    # ============= EVENT CALLBACKS =============

    def on_frame_update(self, callback: FrameUpdateCallback) -> None:
        """Set callback for frame updates"""
        self.on_frame_update_callback = callback

    def on_video_config_update(self, callback: VideoConfigUpdateCallback) -> None:
        """Set callback for video config updates"""
        self.on_video_config_update_callback = callback

    def on_stream_started(self, callback: StreamStartedCallback) -> None:
        """Set callback for stream started events"""
        self.on_stream_started_callback = callback

    def on_stream_stopped(self, callback: StreamStoppedCallback) -> None:
        """Set callback for stream stopped events"""
        self.on_stream_stopped_callback = callback

    def on_recovery_triggered(self, callback: RecoveryTriggeredCallback) -> None:
        """Set callback for recovery triggered events"""
        self.on_recovery_triggered_callback = callback

    def on_status_update(self, callback: StatusUpdateCallback) -> None:
        """Set callback for status updates"""
        self.on_status_update_callback = callback

    def on_stream_stats(self, callback: StreamStatsCallback) -> None:
        """Set callback for stream statistics"""
        self.on_stream_stats_callback = callback

    # ============= MESSAGE HANDLING =============

    async def _handle_role_specific_message(self, data: dict[str, Any]) -> None:
        """Handle consumer-specific messages"""
        msg_type = data.get("type")

        if msg_type == "frame_update":
            await self._handle_frame_update(data)
        elif msg_type == "video_config_update":
            await self._handle_video_config_update(data)
        elif msg_type == "stream_started":
            await self._handle_stream_started(data)
        elif msg_type == "stream_stopped":
            await self._handle_stream_stopped(data)
        elif msg_type == "recovery_triggered":
            await self._handle_recovery_triggered(data)
        elif msg_type == "status_update":
            await self._handle_status_update(data)
        elif msg_type == "stream_stats":
            await self._handle_stream_stats(data)
        elif msg_type == "participant_joined":
            logger.info(
                f"📥 Participant joined: {data.get('participant_id')} as {data.get('role')}"
            )
            # If it's a producer joining, we should be ready for offers
            if data.get("role") == "producer":
                producer_id = data.get("participant_id", "")
                logger.info(
                    f"🎬 Producer {producer_id} joined - ready for WebRTC offers"
                )
        elif msg_type == "participant_left":
            logger.info(
                f"📤 Participant left: {data.get('participant_id')} ({data.get('role')})"
            )
            # If it's a producer leaving, we should be ready for recovery
            if data.get("role") == "producer":
                producer_id = data.get("participant_id", "")
                logger.info(
                    f"👋 Producer {producer_id} left - waiting for reconnection..."
                )
                # Reset state for potential reconnection
                self.has_remote_description = False
                self.ice_candidate_queue = []
        elif msg_type == "webrtc_offer":
            await self.handle_webrtc_offer(
                data.get("offer", {}), data.get("from_producer", "")
            )
        elif msg_type == "webrtc_answer":
            logger.info("Received WebRTC answer (consumer should not receive this)")
        elif msg_type == "webrtc_ice":
            await self.handle_webrtc_ice(
                data.get("candidate", {}), data.get("from_producer", "")
            )
        elif msg_type == "emergency_stop":
            logger.warning(f"Emergency stop: {data.get('reason', 'Unknown reason')}")
            if self.on_error_callback:
                self.on_error_callback(
                    f"Emergency stop: {data.get('reason', 'Unknown reason')}"
                )
        else:
            logger.warning(f"Unknown message type for consumer: {msg_type}")

    async def _handle_frame_update(self, data: dict[str, Any]) -> None:
        """Handle frame update message"""
        if self.on_frame_update_callback:
            frame_data = FrameData(
                data=data.get("data", b""), metadata=data.get("metadata", {})
            )
            self.on_frame_update_callback(frame_data)

    async def _handle_video_config_update(self, data: dict[str, Any]) -> None:
        """Handle video config update message"""
        if self.on_video_config_update_callback:
            config = self._dict_to_video_config(data.get("config", {}))
            self.on_video_config_update_callback(config)

    async def _handle_stream_started(self, data: dict[str, Any]) -> None:
        """Handle stream started message"""
        if self.on_stream_started_callback:
            config = self._dict_to_video_config(data.get("config", {}))
            participant_id = data.get("participant_id", "")
            self.on_stream_started_callback(config, participant_id)

        logger.info(
            f"🚀 Stream started by producer {data.get('participant_id')}, ready to receive video"
        )

    async def _handle_stream_stopped(self, data: dict[str, Any]) -> None:
        """Handle stream stopped message"""
        producer_id = data.get("participant_id", "")
        reason = data.get("reason")

        logger.info(f"⏹️ Stream stopped by producer {producer_id}")
        if reason:
            logger.info(f"   Reason: {reason}")

        # Reset WebRTC state for potential restart
        self.has_remote_description = False
        self.ice_candidate_queue = []

        # Keep peer connection alive for potential restart
        logger.info("🔄 Ready for stream restart...")

        if self.on_stream_stopped_callback:
            self.on_stream_stopped_callback(producer_id, reason)

    async def _handle_recovery_triggered(self, data: dict[str, Any]) -> None:
        """Handle recovery triggered message"""
        if self.on_recovery_triggered_callback:
            from .types import RecoveryPolicy

            policy = RecoveryPolicy(data.get("policy", "freeze_last_frame"))
            reason = data.get("reason", "")
            self.on_recovery_triggered_callback(policy, reason)

    async def _handle_status_update(self, data: dict[str, Any]) -> None:
        """Handle status update message"""
        if self.on_status_update_callback:
            status = data.get("status", "")
            status_data = data.get("data")
            self.on_status_update_callback(status, status_data)

    async def _handle_stream_stats(self, data: dict[str, Any]) -> None:
        """Handle stream stats message"""
        if self.on_stream_stats_callback:
            from .types import StreamStats

            stats_data = data.get("stats", {})
            stats = StreamStats(
                stream_id=stats_data.get("stream_id", ""),
                duration_seconds=stats_data.get("duration_seconds", 0.0),
                frame_count=stats_data.get("frame_count", 0),
                total_bytes=stats_data.get("total_bytes", 0),
                average_fps=stats_data.get("average_fps", 0.0),
                average_bitrate=stats_data.get("average_bitrate", 0.0),
            )
            self.on_stream_stats_callback(stats)

    # ============= TRACK HANDLING =============

    def _handle_track_received(self, track: Any) -> None:
        """Handle received video track"""
        logger.info(f"📺 Received video track: {track.kind}")
        self.remote_stream = track

        # Start reading frames from the track
        if track.kind == "video":
            asyncio.create_task(self._read_video_frames(track))
            # Start frame monitoring
            asyncio.create_task(self._start_frame_monitoring())

    async def _read_video_frames(self, track: Any) -> None:
        """Read frames from video track and trigger callbacks"""
        frame_count = 0
        self._monitoring_frames = True
        consecutive_errors = 0
        max_consecutive_errors = 5

        try:
            logger.info(f"📹 Starting frame reading from track: {track.kind}")

            while self._monitoring_frames:
                try:
                    # Use timeout to detect stream issues
                    frame = await asyncio.wait_for(track.recv(), timeout=5.0)
                    frame_count += 1
                    consecutive_errors = 0  # Reset error count on success

                    # Update frame monitoring
                    import time

                    self._last_frame_time = time.time()

                    # Convert frame to numpy array properly - use RGB format to match server
                    img = frame.to_ndarray(format="rgb24")

                    # Convert RGB to BGR for OpenCV compatibility if needed
                    # For callbacks, we can provide RGB data and let user decide format
                    frame_data = FrameData(
                        data=img.tobytes(),
                        metadata={
                            "width": frame.width,
                            "height": frame.height,
                            "format": "rgb24",  # Server sends RGB format
                            "pts": frame.pts,
                            "time_base": str(frame.time_base),
                            "frame_count": frame_count,
                        },
                    )

                    # Trigger frame update callback
                    if self.on_frame_update_callback:
                        self.on_frame_update_callback(frame_data)

                    if frame_count % 30 == 0:  # Log every 30 frames
                        logger.info(f"📹 Processed {frame_count} video frames")

                except TimeoutError:
                    logger.warning(
                        "⏰ Timeout waiting for video frame - stream may have stopped"
                    )
                    consecutive_errors += 1
                    if consecutive_errors >= max_consecutive_errors:
                        logger.error(
                            "❌ Too many consecutive frame timeouts - stopping frame reading"
                        )
                        break
                    await asyncio.sleep(1)  # Wait before retrying
                    continue

                except Exception as frame_error:
                    # Individual frame processing error - log but continue
                    consecutive_errors += 1
                    logger.warning(
                        f"⚠️ Error processing frame {frame_count}: {frame_error}"
                    )

                    if consecutive_errors >= max_consecutive_errors:
                        logger.error(
                            f"❌ Too many consecutive frame errors ({consecutive_errors}) - stopping frame reading"
                        )
                        break

                    await asyncio.sleep(0.1)  # Brief pause before retrying
                    continue

        except Exception as e:
            logger.error(f"❌ Fatal error in frame reading loop: {e}")

        finally:
            logger.info(
                f"📊 Frame reading stopped. Total frames processed: {frame_count}"
            )
            self._monitoring_frames = False

            # If we stopped due to errors and we're still connected, try to restart
            if consecutive_errors >= max_consecutive_errors and self.connected:
                logger.info(
                    "🔄 Frame reading stopped due to errors - triggering connection recovery"
                )
                asyncio.create_task(self._handle_connection_failure())

    async def _start_frame_monitoring(self) -> None:
        """Start monitoring for frame timeouts"""
        if self._frame_timeout_task and not self._frame_timeout_task.done():
            self._frame_timeout_task.cancel()

        self._frame_timeout_task = asyncio.create_task(self._monitor_frame_timeout())

    async def _monitor_frame_timeout(self) -> None:
        """Monitor for frame timeouts and trigger recovery if needed"""
        timeout_seconds = 10.0  # No frames for 10 seconds = problem

        while self.connected and self._monitoring_frames:
            await asyncio.sleep(5)  # Check every 5 seconds

            if self._last_frame_time is not None:
                import time

                time_since_last_frame = time.time() - self._last_frame_time

                if time_since_last_frame > timeout_seconds:
                    logger.warning(
                        f"⚠️ No frames received for {time_since_last_frame:.1f}s - stream may be stopped"
                    )
                    # Reset frame monitoring
                    self._last_frame_time = None

    # ============= UTILITY METHODS =============

    @staticmethod
    async def create_and_connect(
        workspace_id: str,
        room_id: str,
        base_url: str = "http://localhost:8000",
        participant_id: str | None = None,
    ) -> "VideoConsumer":
        """Create a consumer and automatically connect to a room"""
        consumer = VideoConsumer(base_url)
        connected = await consumer.connect(workspace_id, room_id, participant_id)

        if not connected:
            raise ValueError("Failed to connect as video consumer")

        return consumer

    def attach_to_video_element(self, video_element: Any) -> None:
        """Attach remote stream to a video element (for web frameworks)"""
        if self.remote_stream:
            # This would be used in web contexts
            # For now, just log that we have a stream
            logger.info("Video stream available for attachment")

    async def get_video_stats(self) -> WebRTCStats | None:
        """Get current video statistics"""
        return await self.get_stats()

    def _dict_to_video_config(self, data: dict[str, Any]) -> VideoConfig:
        """Convert dictionary to VideoConfig"""
        from .types import Resolution, VideoEncoding

        config = VideoConfig()

        if "encoding" in data:
            config.encoding = VideoEncoding(data["encoding"])

        if "resolution" in data:
            res_data = data["resolution"]
            config.resolution = Resolution(
                width=res_data.get("width", 640), height=res_data.get("height", 480)
            )

        if "framerate" in data:
            config.framerate = data["framerate"]

        if "bitrate" in data:
            config.bitrate = data["bitrate"]

        if "quality" in data:
            config.quality = data["quality"]

        return config
